### PR TITLE
[Order] Do not stop cancelling unpaid orders on exception during canc…

### LINKED ISSUE #11768: sylius:cancel-unpaid-orders halts if exception thrown on single order
**Sylius version affected**: 1.3.x and up

The command `sylius:cancel-unpaid-orders` is being used in our shop regularly (cron scheduled) to cancel old orders. In one of the event listeners to this event, we send the customer an email. All fine, but I just noticed someone entered `gmail..com` in their email address. This results in an exception on the sending of that email, which occurs as soon as the state machine kicks in to cancel the order.

Now one could reason that the email sending logic should handle this (the event listener in this case), which makes full sense. But next to this, any exception that happens inside a single order will make this command stop instantly and not process the other orders. This can keep blocking forever, making the list of unpaid orders big and they never get cancelled. 

From that perspective it might be good to catch exceptions separately and let the cancel updater continue to the next order.

### LINKED ISSUE #11945: Cancel unpaid orders command should process next orders upon error
**Describe the proposed solution**
The command "sylius:cancel-unpaid-orders" should be atomic on each order cancellation, so that if an order fails to be canceled, andit throws an error, next orders can still be processed for cancellation. This could be done easily with a try/catch and a writeln to output or a message in the log.

### PR DESCRIPTION
**Pull request by twojtylak**
…elling a single order

| Q               | A
| --------------- | -----
| Branch?         | 1.7
| Bug fix?        | yes
| New feature?    | no
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | fixes #11768 & fixes #11945
| License         | MIT

### CODE REVIEW

File: src/Sylius/Component/Core/Updater/UnpaidOrdersStateUpdater.php [hunk 1]
```diff
@@ -29,24 +30,36 @@ final class UnpaidOrdersStateUpdater implements UnpaidOrdersStateUpdaterInterfac
...
+
     /**
      * @param string $expirationPeriod
      */
     public function __construct(
         OrderRepositoryInterface $orderRepository,
         Factory $stateMachineFactory,
-        $expirationPeriod
+        $expirationPeriod,
+        LoggerInterface $logger
```
Comment (loevgaard) [inline src/Sylius/Component/Core/Updater/UnpaidOrdersStateUpdater.php:43]:
```suggestion
        LoggerInterface $logger = null
```

You need to allow this to be null, else it's a BC break and can't be in 1.x. You could (should) trigger a deprecation warning if this is null

File: src/Sylius/Bundle/CoreBundle/Resources/config/services.xml [hunk 2]
```diff
@@ -186,6 +186,7 @@
             <argument type="service" id="sylius.repository.order" />
             <argument type="service" id="sm.factory" />
             <argument>%sylius_order.order_expiration_period%</argument>
+            <argument type="service" id="monolog.logger" />
```
Comment (loevgaard) [inline src/Sylius/Bundle/CoreBundle/Resources/config/services.xml:189]: Isn't there a service just called `logger`?

**Review by loevgaard — CHANGES_REQUESTED** [review]

**Comment by loevgaard** [comment]
Also please add to your PR description: `Fixes #11945`

File: src/Sylius/Component/Core/Updater/UnpaidOrdersStateUpdater.php (hunk 1 above)
Comment (twojtylak) [reply src/Sylius/Component/Core/Updater/UnpaidOrdersStateUpdater.php]: Done

File: src/Sylius/Bundle/CoreBundle/Resources/config/services.xml (hunk 2 above)
Comment (twojtylak) [reply src/Sylius/Bundle/CoreBundle/Resources/config/services.xml]: Done. `logger` is an alias for "monolog.logger", so I think it wont make any difference

**Comment by loevgaard** [comment]
Your `Fixes` statement only closes 11768 right now. I think you should prepend `Fixes` to each issue number to have it closed :) Just a heads up ;)

**Comment by twojtylak** [comment]
> Your `Fixes` statement only closes 11768 right now. I think you should prepend `Fixes` to each issue number to have it closed :) Just a heads up ;)

Alright I added `fixes` in the pr description 😄

File: src/Sylius/Component/Core/Updater/UnpaidOrdersStateUpdater.php [hunk 3]
```diff
@@ -29,24 +30,43 @@ final class UnpaidOrdersStateUpdater implements UnpaidOrdersStateUpdaterInterfac
...
 
     public function cancel(): void
     {
         $expiredUnpaidOrders = $this->orderRepository->findOrdersUnpaidSince(new \DateTime('-' . $this->expirationPeriod));
         foreach ($expiredUnpaidOrders as $expiredUnpaidOrder) {
-            $this->cancelOrder($expiredUnpaidOrder);
+            try {
+                $this->cancelOrder($expiredUnpaidOrder);
+            } catch (\Exception $e) {
+                $this->logger && $this->logger->error(
```
Comment (lchrusciel) [inline src/Sylius/Component/Core/Updater/UnpaidOrdersStateUpdater.php:65]: Can we make explicit if statement to improve readability?

File: src/Sylius/Component/Core/spec/Updater/UnpaidOrdersStateUpdaterSpec.php [hunk 4]
```diff
@@ -24,9 +25,9 @@
 
 final class UnpaidOrdersStateUpdaterSpec extends ObjectBehavior
 {
-    function let(OrderRepositoryInterface $orderRepository, Factory $stateMachineFactory): void
+    function let(OrderRepositoryInterface $orderRepository, Factory $stateMachineFactory, LoggerInterface $logger): void
     {
-        $this->beConstructedWith($orderRepository, $stateMachineFactory, '10 months');
+        $this->beConstructedWith($orderRepository, $stateMachineFactory, '10 months', $logger);
```
Comment (lchrusciel) [inline src/Sylius/Component/Core/spec/Updater/UnpaidOrdersStateUpdaterSpec.php:30]: Can we make a unit test for behavior without logger?

**Review by lchrusciel — COMMENTED** [review]
Hey @twojtylak,

thanks for the contribution. I think it would be worthy to have it merged in the Core. Can you make two more improvements in your code?

File: src/Sylius/Component/Core/Updater/UnpaidOrdersStateUpdater.php (hunk 3 above)
Comment (twojtylak) [reply src/Sylius/Component/Core/Updater/UnpaidOrdersStateUpdater.php]: Hey @lchrusciel,

I just added an explicit `instance of` check.

File: src/Sylius/Component/Core/spec/Updater/UnpaidOrdersStateUpdaterSpec.php (hunk 4 above)
Comment (twojtylak) [reply src/Sylius/Component/Core/spec/Updater/UnpaidOrdersStateUpdaterSpec.php:30]: Hey @lchrusciel,

I added the `it_wont_stop_cancelling_unpaid_orders_on_exception_for_a_single_order_and_skips_logging_if_logger_is_not_set` (I am not sure if the name is meaningful enough or too long) test. I hope that's okay. Cheers

File: src/Sylius/Component/Core/Updater/UnpaidOrdersStateUpdater.php [hunk 5]
```diff
@@ -29,24 +30,45 @@ final class UnpaidOrdersStateUpdater implements UnpaidOrdersStateUpdaterInterfac
...
 
     public function cancel(): void
     {
         $expiredUnpaidOrders = $this->orderRepository->findOrdersUnpaidSince(new \DateTime('-' . $this->expirationPeriod));
         foreach ($expiredUnpaidOrders as $expiredUnpaidOrder) {
-            $this->cancelOrder($expiredUnpaidOrder);
+            try {
+                $this->cancelOrder($expiredUnpaidOrder);
+            } catch (\Exception $e) {
+                if ($this->logger instanceof LoggerInterface) {
```
Comment (loevgaard) [inline src/Sylius/Component/Core/Updater/UnpaidOrdersStateUpdater.php:65]:
```suggestion
                if (null !== $this->logger) {
```

File: src/Sylius/Component/Core/spec/Updater/UnpaidOrdersStateUpdaterSpec.php (hunk 4 above)
Comment (lchrusciel) [reply src/Sylius/Component/Core/spec/Updater/UnpaidOrdersStateUpdaterSpec.php:30]: Hey @twojtylak, it is understandable enough ;)

**Comment by lchrusciel** [comment]
Hey @twojtylak, I've restarted also your build.

**Review by loevgaard — APPROVED** [review]

**Comment by lchrusciel** [comment]
Thanks, @twojtylak! :1st_place_medal:

**Comment by LucaGallinari** [comment]
Thanks!

**Pull request merged**